python/taichi/lang/impl.py

```diff
@@ -813,6 +813,12 @@ def ti_format_list_to_content_entries(raw):
     def entry2content(_var):
         if isinstance(_var, str):
             return _var
+        # handle optional format specifier
+        if isinstance(_var, list):
+            assert len(_var) == 2 and (isinstance(_var[1], str)
+                                       or _var[1] is None)
+            _var[0] = Expr(_var[0]).ptr
+            return _var
         return Expr(_var).ptr
```

Comment:
Let's try to make the return from different branches as consistent as possible, otherwise it's gonna pollute the code that make use of the function result (`extract_formats` in your case). For example, always return a tuple of "content, format" even if the format is left with None

Reply:
get~